File: 이상욱/BOJ/m3w11/BOJ22871.java
package m3w11;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ22871 {
	static int N;
	static long[] arr;
	static long[] dp;
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		N = Integer.parseInt(br.readLine());
		arr = new long[N]; 
		dp = new long[N];
		
		StringTokenizer st = new StringTokenizer(br.readLine());
		for(int i = 0; i < N; i++) {
			arr[i] = Long.parseLong(st.nextToken());
			dp[i] = -1;
		}
		
		System.out.println(cross(0));
		
	}
	private static long cross(int x) {
		
		if( x == N-1 ) {
			return 0;
		}
		
		if(dp[x] != -1) {
			return dp[x];
		}
		
		dp[x] = Long.MAX_VALUE;
		
		for(int i  = x+1; i < N; i++) {
			dp[x] = Math.min(dp[x], Math.max(cross(i), (i-x) * (1 + Math.abs(arr[x] - arr[i]))));
		}
		
		return dp[x];
	}
}

File: 이상욱/BOJ/m3w11/BOJ22871.kt
package m3w11

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

var n = 0
lateinit var arr : LongArray
lateinit var dp : LongArray
fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))

    n = br.readLine().toInt()
    arr = LongArray(n)
    dp = LongArray(n)

    val st = StringTokenizer(br.readLine())
    repeat(n){ i ->
        arr[i] = st.nextToken().toLong()
        dp[i] = -1
    }

    println(cross(0))
    

}

fun cross(x: Int): Long {

    if( x == n-1){
        return 0
    }

    if(dp[x] != -1L){
        return dp[x]
    }

    dp[x] = Long.MAX_VALUE

    for(i in x+1 until n){
        dp[x] = Math.min(dp[x] , Math.max(cross(i), (i - x) * (1 + Math.abs(arr[x] - arr[i]))))
    }
    return dp[x]

}
